File: tests/ui/crud/update-share.js
const {chromium} = require('playwright')
const path = require('path')
const {
  login,
  getConfig,
  getWalletsFromStorage
} = require('../auth-helper')

/**
 * Test: Update wallet share permissions
 * This test updates permissions on an existing wallet share
 */

;(async () => {
  const browser = await chromium.launch({headless: true, slowMo: 500})
  const page = await browser.newPage()
  const config = getConfig()

  let success = false

  try {
    console.log('🚀 Starting update wallet share test...')

    // Step 1: Login as admin
    console.log('📝 Step 1: Logging in as admin...')
    await login(page)

    // Step 2: Check user has a wallet via UI
    console.log('📝 Step 2: Checking user has a wallet...')
    await page.goto(`${config.baseUrl}/wallet`)
    await page.waitForTimeout(2000)

    // Count wallet items in UI
    const walletCount = await page
      .locator('.q-drawer .q-list .q-item')
      .filter({hasNot: page.locator('text=Add a new wallet')})
      .count()

    if (walletCount === 0) {
      throw new Error(
        'No wallets found in UI. Please run create-wallet.js first.'
      )
    }

    console.log(`✅ User has ${walletCount} wallet(s)`)

    // Get wallet data from localStorage for API calls
    const wallets = await getWalletsFromStorage(page)
    const existingWallet = wallets[0]
    const walletId = existingWallet.id
    const adminKey = existingWallet.adminkey

    console.log(`✅ Using wallet ID: ${walletId}`)

    // Step 3: Get existing shares
    console.log('📝 Step 3: Getting existing shares...')
    const response = await page.request.get(
      `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
      {
        headers: {'X-Api-Key': adminKey}
      }
    )

    if (!response.ok()) {
      throw new Error(`Failed to get shares: ${response.status()}`)
    }

    let shares = await response.json()

    // If no shares exist, create one first
    if (!Array.isArray(shares) || shares.length === 0) {
      console.log('⚠️ No shares found. Creating one first...')
      const shareWithUser = config.secondaryUsername
      if (!shareWithUser) {
        throw new Error('LNBITS_SECONDARY_USERNAME must be set in .env.local')
      }

      const createResponse = await page.request.post(
        `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
        {
          headers: {
            'X-Api-Key': adminKey,
            'Content-Type': 'application/json'
          },
          data: JSON.stringify({
            user_id: shareWithUser,
            permissions: 1
          })
        }
      )

      if (!createResponse.ok()) {
        throw new Error(`Failed to create share: ${createResponse.status()}`)
      }

      console.log('✅ Created test share')

      // Re-fetch shares
      const refetchResponse = await page.request.get(
        `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
        {
          headers: {'X-Api-Key': adminKey}
        }
      )
      shares = await refetchResponse.json()
    }

    const shareToUpdate = shares[0]
    console.log(`📄 Found share to update: ${shareToUpdate.id}`)
    console.log(`   Current permissions: ${shareToUpdate.permissions}`)

    // Step 4: Update permissions via API
    console.log('📝 Step 4: Updating share permissions...')

    // Change permissions (if VIEW=1, change to VIEW+CREATE_INVOICE=3)
    const newPermissions = shareToUpdate.permissions === 1 ? 3 : 1

    const updateResponse = await page.request.put(
      `${config.baseUrl}/api/v1/wallet_shares/${shareToUpdate.id}`,
      {
        headers: {
          'X-Api-Key': adminKey,
          'Content-Type': 'application/json'
        },
        data: JSON.stringify({permissions: newPermissions})
      }
    )

    if (updateResponse.ok()) {
      const updatedShare = await updateResponse.json()
      console.log(`✅ Share updated successfully`)
      console.log(`   New permissions: ${updatedShare.permissions}`)
      success = true
    } else {
      console.log(`❌ Update failed: ${updateResponse.status()}`)
      const errorText = await updateResponse.text()
      console.log(`   Error: ${errorText}`)
    }

    // Step 5: Verify update
    console.log('📝 Step 5: Verifying update...')
    const verifyResponse = await page.request.get(
      `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
      {
        headers: {'X-Api-Key': adminKey}
      }
    )

    if (verifyResponse.ok()) {
      const updatedShares = await verifyResponse.json()
      const verifiedShare = updatedShares.find(s => s.id === shareToUpdate.id)

      if (verifiedShare && verifiedShare.permissions === newPermissions) {
        console.log('✅ Update verified - permissions match!')
        success = true
      } else {
        console.log('❌ Update verification failed - permissions do not match')
        success = false
      }
    }

    // Take screenshot
    await page.screenshot({
      path: path.join(__dirname, '../screenshots/update-share-result.png'),
      fullPage: true
    })

    if (success) {
      console.log('🎉 SUCCESS! Share updated successfully!')
    }
  } catch (error) {
    console.error('💥 Error:', error.message)
    await page.screenshot({
      path: path.join(__dirname, '../screenshots/update-share-error.png'),
      fullPage: true
    })
  } finally {
    await browser.close()
  }

  process.exit(success ? 0 : 1)
})()

File: tests/api/update_share.py
#!/usr/bin/env python3
"""
Test: Update wallet share permissions via API
This test updates permissions on an existing wallet share using the REST API
"""

import asyncio
import os

import httpx
from loguru import logger


# Load config from .env.local
def load_config():
    config = {}
    env_path = os.path.join(os.path.dirname(__file__), "../../.env.local")

    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, value = line.split("=", 1)
                    config[key.strip()] = value.strip()

    return {
        "base_url": config.get("TEST_LNBITS_URL", os.getenv("TEST_LNBITS_URL")),
        "admin_key": config.get("TEST_ADMIN_API_KEY", os.getenv("TEST_ADMIN_API_KEY")),
        "wallet_id": config.get("TEST_WALLET_ID", os.getenv("TEST_WALLET_ID")),
        "username": config.get(
            "LNBITS_ADMIN_USERNAME", os.getenv("LNBITS_ADMIN_USERNAME")
        ),
        "password": config.get(
            "LNBITS_ADMIN_PASSWORD", os.getenv("LNBITS_ADMIN_PASSWORD")
        ),
    }


async def test_update_share():
    """Test updating wallet share permissions"""
    config = load_config()

    if not config["admin_key"]:
        logger.error("❌ TEST_ADMIN_API_KEY must be set in .env.local")
        return False

    if not config["wallet_id"]:
        logger.error("❌ TEST_WALLET_ID must be set in .env.local")
        return False

    logger.info("🚀 Starting update wallet share API test...")
    logger.info(f"📍 Base URL: {config['base_url']}")
    logger.info(f"📍 Wallet ID: {config['wallet_id']}")

    async with httpx.AsyncClient() as client:
        # Step 1: Get existing shares
        logger.info("📝 Step 1: Getting existing shares...")
        response = await client.get(
            f"{config['base_url']}/api/v1/wallet_shares/{config['wallet_id']}",
            headers={"X-Api-Key": config["admin_key"]},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to get shares: {response.status_code}")
            return False

        shares = response.json()

        if not isinstance(shares, list) or len(shares) == 0:
            logger.warning("⚠️  No shares found to update. Please create a share first.")
            return True  # Not a failure, just nothing to update

        share_to_update = shares[0]
        logger.info(f"📄 Found share to update: {share_to_update['id']}")
        logger.info(f"   Current permissions: {share_to_update['permissions']}")

        # Step 2: Update permissions (using admin key)
        logger.info("📝 Step 2: Updating share permissions...")

        # Toggle between VIEW (1) and VIEW+CREATE_INVOICE (3)
        new_permissions = 3 if share_to_update["permissions"] == 1 else 1

        response = await client.put(
            f"{config['base_url']}/api/v1/wallet_shares/{share_to_update['id']}",
            headers={
                "X-Api-Key": config["admin_key"],
                "Content-Type": "application/json",
            },
            json={"permissions": new_permissions},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to update share: {response.status_code}")
            logger.error(f"   Response: {response.text}")
            return False

        updated_share = response.json()
        logger.info("✅ Share updated successfully!")
        logger.info(f"   New permissions: {updated_share['permissions']}")

        # Step 3: Verify update
        logger.info("📝 Step 3: Verifying update...")
        response = await client.get(
            f"{config['base_url']}/api/v1/wallet_shares/{config['wallet_id']}",
            headers={"X-Api-Key": config["admin_key"]},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to verify: {response.status_code}")
            return False

        updated_shares = response.json()
        verified_share = next(
            (s for s in updated_shares if s["id"] == share_to_update["id"]), None
        )

        if verified_share and verified_share["permissions"] == new_permissions:
            logger.info("✅ Update verified - permissions match!")
            logger.info("\n🎉 SUCCESS! Share updated successfully!")
            return True
        else:
            logger.error("❌ Update verification failed - permissions do not match")
            return False


if __name__ == "__main__":
    success = asyncio.run(test_update_share())
    exit(0 if success else 1)
